### PR TITLE
[QT-576] Optimize build workflow

### PR DESCRIPTION
**Pull request by ryancragun**
Improve our build workflow execution time by using custom runners,
improved caching and conditional Web UI builds.

Runners
-------
We improve our build times[0] by using larger custom runners[1] when
building the UI and Vault.

Caching
-------
We improve Vault caching by keeping a cache for each build job. This
strategy has the following properties which should result in faster
build times when `go.sum` hasn't been changed from prior builds, or
when a pull request is retried or updated after a prior successful
build:

* Builds will restore cached Go modules and Go build cache according to
  the Go version, platform, architecture, go tags, and hash of `go.sum`
  that relates to each individual build workflow. This reduces the
  amount of time it will take to download the cache on hits and upload
  the cache on misses.
* Parallel build workflows won't clobber each others build cache. This
  results in much faster compile times after cache hits because the Go
  compiler can reuse the platform, architecture, and tag specific build
  cache that it created on prior runs.
* Older modules and build cache will not be uploaded when creating a new
  cache. This should result in lean cache sizes on an ongoing basis.
* On cache misses we will have to upload our compressed module and build
  cache. This will slightly extend the build time for pull requests that
  modify `go.sum`.

Web UI
------
We no longer build the web UI in every build workflow. Instead we separate
the UI building into its own workflow and cache the resulting assets.
The same UI assets are restored from cache during build worklows. This
strategy has the following properties:

* If the `ui` directory has not changed from prior builds we'll restore
  `http/web_ui` from cache and skip building the UI for no reason.
* We continue to use the built-in `yarn` caching functionality in
  `action/setup-node`. The default mode saves the `yarn` global cache.
  to improve UI build times if the cache has not been modified.

Changes
-------
* Add per platform/archicture Go module and build caching
* Move UI building into a separate job and cache the result
* Restore UI cache during build
* Pin workflows

Notes
-----
[0] https://hashicorp.atlassian.net/browse/QT-578
[1] https://github.com/hashicorp/vault/actions/runs/5415830307/jobs/9844829929

### CODE REVIEW

**Comment by ryancragun** [comment]
Example of a run where we get cache hits for both Go and the UI https://github.com/hashicorp/vault/actions/runs/5394083885

File: .github/workflows/build-vault-oss.yml [hunk 1]
```diff
@@ -40,18 +49,31 @@ jobs:
...
+          cache: false # Use our own caching strategy for better cross platform support
+      - name: Set up Go module cache
+        uses: actions/cache@88522ab9f39a2ea568f7027eddc7d8d8bc9d59c8 # v3.3.1
+        with:
+          path: |
+            ${{ inputs.go-cache }}
+            ${{ inputs.go-mod-cache }}
+          # Manage the cache for each platform and architecture combo. In this way we'll only ever
+          # restore modules and build cache for the platform and architecture we're targeting.
+          key: ${{ inputs.go-version }}-${{ inputs.goos }}-${{ inputs.goarch }}-go-${{ hashFiles('**/go.sum') }}
```
Comment (ryancragun) [inline .github/workflows/build-vault-oss.yml:65]: We should probably include a hash of `go-tags` in here as well. In the event we build for the same platform/architecture but with different tags, eg: ent, hsm, fips, hsm.fips, we'd want to have separate build caches.

File: .github/workflows/build-vault-oss.yml [hunk 2]
```diff
@@ -40,18 +49,34 @@ jobs:
...
+      - uses: actions/setup-go@fac708d6674e30b6ba41289acaab6d4b75aa0753 # v4.0.1
         with:
           go-version: ${{ inputs.go-version }}
-      - name: Set up node and yarn
-        uses: actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c # v3.6.0
+          cache: false # Use our own caching strategy for better cross platform support
+      - name: Set up Go cache key tags
+        id: cache-key-tags
+        run: echo "gotags=$(echo ${{ inputs.go-tags }} | tr ' ' '-')" >> "$GITHUB_ENV"
+      - name: Set up Go cache
```
Comment (ncabatoff) [inline .github/workflows/build-vault-oss.yml:60]: Any reason not to do this in ci.yml too?

**Comment by ryancragun** [comment]
When using larger instances with cache hits time was reduced to 12 minutes https://github.com/hashicorp/vault/actions/runs/5426525055

File: .github/workflows/drepecated-functions-checker.yml [hunk 3]
```diff
@@ -12,11 +12,11 @@ jobs:
     timeout-minutes: 30
```
Comment (rebwill) [inline .github/workflows/drepecated-functions-checker.yml:12]: Nit: this file name should be `deprecated-functions-checker.yml`.

**Review by rebwill — APPROVED** [review]

File: .github/workflows/build-vault-oss.yml (hunk 2 above)
Comment (ryancragun) [reply .github/workflows/build-vault-oss.yml:60]: Definitely makes sense to me. We could probably use newer instance classes for the CI workflows as well. They're more performant and cheaper.

**Pull request merged**